File: RTIOW/chap3/main.py
#!/bin/env python
# -*- coding: utf-8 -*-

"""
A simple program to write a PPM file.

From chapter 3 of "Ray Tracing in One Weekend".
Uses the Ray and Vec3 classes.
"""

import vec3
import ray

def color(r):
    """Get colour from vector."""

    unit_direction = r.direction.unit_vector()
    t = (unit_direction.y + 1.0) * 0.5
    return vec3.Vec3((1.0, 1.0, 1.0)) * (1.0 - t) + vec3.Vec3((0.5, 0.7, 1.0)) * t

nx = 200
ny = 100

print('P3\n%d %d 255' % (nx, ny))

lower_left_corner = vec3.Vec3((-2.0, -1.0, -1.0))
horizontal = vec3.Vec3((4.0, 0.0, 0.0))
vertical = vec3.Vec3((0.0, 2.0, 0.0))
origin = vec3.Vec3((0.0, 0.0, 0.))

for j in range(ny)[::-1]:
    for i in range(nx):
        u = float(i) / float(nx)
        v = float(j) / float(ny)
        r = ray.Ray(origin, lower_left_corner + horizontal * u + vertical * v)
        col = color(r)
        ir = int(255.99 * col.r)
        ig = int(255.99 * col.g)
        ib = int(255.99 * col.b)

        print('%d %d %d' % (ir, ig, ib))

File: RTIOW/python/chap3/main.c
/*
 * A simple program to write a PPM file.
 * 
 * From chapter 3 of "Ray Tracing in One Weekend".
 * Uses the Ray and Vec3 modules.
 */

#include <stdio.h>

#include "vec3.h"
#include "ray.h"


vec3 *
color(ray *r)
{
    vec3 *unit_direction = vec3_unit_vector(ray_direction(r));
    float t = 0.5 * (vec3_y(unit_direction) + 1.0);

    vec3 *rhs = vec3_mul_float(vec3_new_vals(0.5, 0.7, 1.0), t);
    vec3 *lhs = vec3_mul_float(vec3_new_vals(1.0, 1.0, 1.0), (1.0 - t));

    return vec3_add_vec3(lhs, rhs);
}

int main(void)
{
    int nx = 200;
    int ny = 100;

    printf("P3\n%d %d 255\n", nx, ny);

    vec3 *lower_left_corner = vec3_new_vals(-2.0, -1.0, -1.0);
    vec3 *horizontal = vec3_new_vals(4.0, 0.0, 0.0);
    vec3 *vertical = vec3_new_vals(0.0, 2.0, 0.0);
    vec3 *origin = vec3_new_vals(0.0, 0.0, 0.0);

    for (int j = ny - 1; j >= 0; --j)
    {
        for (int i = 0; i < nx; ++i)
        {
            float u = (float) i / (float) nx;
            float v = (float) j / (float) ny;
            vec3 *accum = vec3_mul_float(vertical, v);
            vec3 *tmp = vec3_mul_float(horizontal, u);
            accum = vec3_add_vec3(accum, tmp);
            accum = vec3_add_vec3(accum, lower_left_corner);
            ray *r = ray_new(origin, accum);
//            printf("r=%s\n", ray_str(r));

            vec3 *col = color(r);
//            printf("col=%s\n", vec3_str(col));
//            printf("vec3_r(col)=%f\n", vec3_r(col));
            int ir = (int) (255.99 * vec3_r(col));
//            printf("ir=%d\n", ir);
            int ig = (int) (255.99 * vec3_g(col));
            int ib = (int) (255.99 * vec3_b(col));

            printf("%d %d %d\n", ir, ig, ib);
        }
    }

    return 0;
}
